gnuradio-runtime/python/gnuradio/gr/gateway.py

```diff
@@ -128,36 +154,47 @@ def __gr_block_handle(self):
         """
         Dispatch tasks according to the action type specified in the message.
         """
+
         if self.__message.action == gr.block_gw_message_type.ACTION_GENERAL_WORK:
+            # Actual number of inputs/output from scheduler
+            ninputs = len(self.__message.general_work_args_input_items)
+            noutputs = len(self.__message.general_work_args_output_items)
+            in_types = self.__in_sig.port_types(ninputs)
+            out_types = self.__out_sig.port_types(noutputs)
             self.__message.general_work_args_return_value = self.general_work(
 
                 input_items=[pointer_to_ndarray(
                     self.__message.general_work_args_input_items[i],
-                    self.__in_sig[i],
+                    in_types[i],
                     self.__message.general_work_args_ninput_items[i]
-                ) for i in self.__in_indexes],
+                ) for i in xrange(ninputs)],
```

Comment:
As much as I like `xrange()` for speed improvements in Py2k this will break in Py3k. Consider using `range()` even if it's a bit slower in Py2k. As this is probably good work we want to have in the `next` branch.

Reply:
range in Python 3 is xrange in Python 2. Since GR 3.7 is Python 2 only, can we keep it xrange here and change it to range for the new branch?

Reply:
Performance difference is trivial in testing - changing to range.